src/wp-includes/query.php

```diff
@@ -1117,15 +1105,40 @@ function _find_post_by_old_slug( $post_type ) {
 		$query .= $wpdb->prepare( ' AND DAYOFMONTH(post_date) = %d', get_query_var( 'day' ) );
 	}
 
+	$key          = md5( $query );
+	$cache_key = "_find_post_by_old_slug:$key";
+	$cache     = wp_cache_get( $cache_key, 'posts' );
+	if ( false !== $cache ) {
+		return $cache;
+	}
+
 	$id = (int) $wpdb->get_var( $query );
 
-	return $id;
+	/**
+	 * Filters the old slug redirect post ID.
+	 *
+	 * @since 4.9.3
+	 *
+	 * @param int $id The redirect post ID.
+	 */
+	$id = apply_filters( 'old_slug_redirect_post_id', $id );
+
+	if ( ! $id ) {
+		return;
+	}
+
+	$link = get_permalink( $id );
+
+	wp_cache_set( $cache_key, $link, 'posts' );
```

Comment:
```suggestion
	wp_cache_set( $cache_key, $id, 'posts' );
```

We should not be caching the link. This value change and is filtered. Let's cache the id instead, it is a smaller cache value.